rfcs/0088-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,156 @@
+concrete rules on due to the difficulty of identifying these breakages. In
+order to be covered by this policy please put tests on your packages.)
+
+# Motivation
+
+Right now the procedure for making breaking changes varies widely by maintainer
+and is largely tribal knowledge. There are a wide variety of approaches ranging
+from:
+
+- After testing to confirm that a change is correct, merge and notify
+  dependents that they need to update their packages.
+- Merge to `staging`, notify dependents then eventually merge to `master`.
+- Gather a large number of fixes in the PR branch then merge to `staging` or
+  `master` once most or all dependencies are fixed.
+
+This adhoc nature has a number of downsides:
+
+- Change authors need to put effort into deciding how to handle each particular
+  change.
+- Maintainers of dependant packages don't know what to expect and have a hard
+  time keeping up with upstream breakages.
+- Merges to `master` or `staging` can result in broken channels which block all
+  updates including security updates.
+
+This aims to provide a uniform approach so that everyone involved knows what to
+expect and to allow further workflows including automation to be build on top
+of these procedures.
+
+## Past Discussions
+
+https://discourse.nixos.org/t/please-fix-all-packages-which-were-broken-by-the-qt-wrapping-changes/6444
+
+# Detailed design
+
+## Goals
+
+The following procedure will be used for making breaking changes. It has a few
+primary goals:
+
+- Avoid putting more burden than necessary on the dependency maintainer. If the
+  maintainers of core derivations face toil proportionally to the number of
+  transitive dependencies they will quickly become overloaded. These
+  maintainers are arguably the most critical to Nixpkgs and  their load needs
+  to be kept manageable.
+- Avoid unnecessarily breaking packages for any period of time. There are a
+  number of users on the `*-unstable` channels and it is annoying if packages
+  in their configuration get broken.
+- Avoid breaking the `master` branch as much as possible. The `master` branch
+  is used by the `*-unstable` channels and breaking this branch means that no
+  updates can be pushed. This includes critical security updates as well as
+  regular feature updates.
+
+## Procedure
+
+The target branch for a merge is not affected by this policy. It will be
+picked, as it is today, according to
+https://nixos.org/manual/nixpkgs/stable/#submitting-changes-commit-policy.
+
+1. The maintainer will prepare a PR with their intended changes.
+2. The maintainer should test a sample of dependent derivations to ensure that
+their change is not unnecessarily or unintentionally breaking. (Example: Ensure
+that a critical output file was not forgotten) Note that sometimes it **is**
+necessary to break all dependent packages, and the maintainer is not required
+to avoid this.
+    1. A sample of at least 100 packages or 3h of build time is recommend as a guideline.
```

Comment:
I'm okay with committers and active contributors to use my server: https://github.com/jonringer/server-configuration . If you're worried about compute. Since I'm not opening this up to the general public, it's not a solution, just a mitigation; but the offer is there.

Obviously, this is guideline is meant "within expectations" of a user's ability. Having some sort `nixpkgs-review` does make it a lot easier for reviewers to merge.

Reply:
This is why we added the time limit and called it a guideline. The idea is even if you only own a potato we hope that you can spare 3h of your potato or find someone else to do so for big changes. If you think 3h is still to muck to ask I'd be happy to consider other thresholds, it was picked mostly arbitrarily to provide a rough target.